framework/src/utils/BilinearInterpolation.C

```diff
@@ -76,32 +88,498 @@ BilinearInterpolation::sample(const T & xcoord, const T & ycoord) const
+      const auto & fQ02 = _zSurface(uy, ux - 1); // new lx at uy
+      const Real & x0 = _x1[ux - 1];             // lx value
+
+      auto dfdx = fQ01 * (y - y2);
+      dfdx += fQ11 * (y2 - y);
+      dfdx += fQ02 * (y1 - y);
+      dfdx += fQ12 * (y - y1);
+      dfdx /= ((x1 - x0) * (y2 - y1));
+      return dfdx;
+    }
+    // Derivative (w/ respect to x) for some point inside box
+    else
+    {
+      auto dfdx_xy = fQ11 * (y - y2);
+      dfdx_xy += fQ21 * (y2 - y);
+      dfdx_xy += fQ12 * (y1 - y);
+      dfdx_xy += fQ22 * (y - y1);
+      dfdx_xy /= ((x2 - x1) * (y2 - y1));
+      return dfdx_xy;
+    }
+  }
+
+  else if (deriv_var == 2)
+  {
+    if (x == x1) // if x equal to x1 node
+    {
+      return (fQ12 - fQ11) / (y2 - y1);
+    }
+    // interior grid line
+    else if ((ly == uy) && ly > 0 && uy < nx2 - 1)
+    {
+      // expand grid size so x1 does not equal x2
+      const auto & fQ10 = _zSurface(ly - 1, lx); // new ly at lx
+      const auto & fQ20 = _zSurface(ly - 1, ux); // new uy at lx
+      const auto & fQ13 = _zSurface(ly + 1, lx); // new ly at ux
+      const auto & fQ23 = _zSurface(ly + 1, ux); // new uy at ux
+      const Real & y0 = _x2[ly - 1];
+      const Real & y3 = _x2[ly + 1];
+
+      auto dfdy_a = fQ10 * (x - x2);
+      dfdy_a += fQ20 * (x1 - x);
+      dfdy_a += fQ11 * (x2 - x);
+      dfdy_a += fQ21 * (x - x1);
+      dfdy_a /= ((x2 - x1) * (y1 - y0));
+
+      auto dfdy_b = fQ11 * (x - x2);
+      dfdy_b += fQ21 * (x1 - x);
+      dfdy_b += fQ13 * (x2 - x);
+      dfdy_b += fQ23 * (x - x1);
+      dfdy_b /= ((x2 - x1) * (y3 - y1));
+      return 0.5 * (dfdy_a + dfdy_b);
+    }
+    // bottom boundary
+    else if ((ly == uy) && ly == 0)
+    {
+      const auto & fQ13 = _zSurface(ly + 1, lx); // new uy at lx
+      const auto & fQ23 = _zSurface(ly + 1, ux); // new uy at ux
+
+      const Real & y3 = _x2[ly + 1]; // lx value
+
+      auto dfdy = fQ11 * (x - x2);
+      dfdy += fQ21 * (x1 - x);
+      dfdy += fQ13 * (x2 - x);
+      dfdy += fQ23 * (x - x1);
+      dfdy /= ((x2 - x1) * (y3 - y1));
+      return dfdy;
+    }
+    // top boundary
+    else if ((ly == uy) && (uy == nx2 - 1))
+    {
+      const auto & fQ10 = _zSurface(ly - 1, lx); // new ly at lx
+      const auto & fQ20 = _zSurface(ly - 1, ux); // new ly at ux
+      const Real & y0 = _x2[ly - 1];             // lx value
+
+      auto dfdy = fQ10 * (x - x2);
+      dfdy += fQ20 * (x1 - x);
+      dfdy += fQ11 * (x2 - x);
+      dfdy += fQ21 * (x - x1);
+      dfdy /= ((x2 - x1) * (y1 - y0));
+      return dfdy;
+    }
+    else
+    {
+      // Derivative (w/ respect to y) for any point inside box
+      auto dfdy_xy = fQ11 * (x - x2);
+      dfdy_xy += fQ21 * (x1 - x);
+      dfdy_xy += fQ12 * (x2 - x);
+      dfdy_xy += fQ22 * (x - x1);
+      dfdy_xy /= ((x2 - x1) * (y2 - y1));
+      return dfdy_xy;
+    }
+  }
+  mooseError("Bilinear interpolation failed to select a case for x1= ", s1, " x2= ", s2);
+}
```

Comment:
This function is massive. Is it possible to simplify this somehow? Also, could this be applied for AD as well?

Reply:
this is to compute derivative so typically we dont use it with AD (unless we want the second derivative)

currently not simplifiable.
if we change from an average slope to a bigger box slope for the "on the grid line" case, then we can move the computation of dfdy etc outside of the ifs, keeping only the index selection in the if statements

Reply:
kk. I'll consider this resolved then